### DISCUSSION TITLE
Potentiometer Override

### DISCUSSION THREAD

**Discussion by GrblGru** [grblHAL/core#214] [Ideas]
Hi Terje,

I have often been asked by users if it is possible to provide an override function with a simple potentiometer. When milling, it very often happens that you want to start slowly to see if the programmed feed is OK. If everything is ok, you can then slowly increase the feed rate. 
Many professional machines have a potentiometer for this purpose with which you can apply the feed rate programmed in the GCode from 0 - 120%. The 0% position is especially helpful. With this you can stop the movement during the work, e.g. to correct a wasteboard clamp. Afterwards you simply turn the potentiometer up again.

The current possibility available in GRBL to influence the feed with the override function can only be operated with buttons on the keyboard or with the mouse on the screen. A potentiometer directly on the machine would be much more practical.
Especially in situations where you have to react quickly.

What do you think ? Is it possible to implement something like this in GrblHAL or is there already something similar ?

**Comment by terjeio** [comment]
>  is there already something similar ?

Yes, the [encoder plugin](https://github.com/grblHAL/Plugin_encoder/) does that. Currently only available for the iMXRT1062 driver though. For zero feed rate the feed hold input should be used.

The main issue with adding support for encoders (or analog pots) is often lack of pins, and also time to add the needed low level driver code. Plenty of opportunities for developers to join the effort to make it happen...

**Reply by dresco** [reply]
> > is there already something similar ?
> 
> Yes, the [encoder plugin](https://github.com/grblHAL/Plugin_encoder/) does that.

For external controls, I have a control panel [plugin](https://github.com/dresco/Plugin_panel) - it supports encoders for jogging & overrides, plus buttons, display etc.. 

There's a Zephyr implementation of the control panel [firmware](https://github.com/dresco/grblPANEL). Both are currently Modbus only, but planning to give these a bit more attention this winter & will be adding CAN bus support.

Also planning to finish an example/reference panel hw design, it's been working fine for me all year in this prototype form, but would be nice to move it on from it's current state..

![image](https://user-images.githubusercontent.com/996823/198816787-ad525658-5c1a-436a-9ed7-7e37272a6881.png)

**Comment by GrblGru** [comment]
@dresco
the panel looks great and I believe you that it leaves nothing to be desired. 

But I am looking for a very simple and cheap method for the override.
There are many young users of my program, who make their first steps with the CNC technology and they can't afford bigger expenses.

@Terje
Of course I would like to contribute something to the project. The only question is if I am able to do so.
Is there somewhere a instruction or an example how to manipulate the override ?
I assume that all boards have an analog input. The problem I see more in adding the value to the override.

**Reply by karoria** [reply]
@GrblGru 
I have already implemented industrial cnc type overrides for feedrate and rpm. I use my own webUI named grblTouch, which is also opensource (hosted on github) and designed for small touchscreen displays. Further, I have implemented shortcut keys for most of the functions/commands. So my setup is, use a 7 inch touch display with rpi as webui client and use an HID keypad (i use pi pico with circuitpython code in it) to fire those keystrokes to rpi. You can easily program any hid compatible microcontroller to send keystrokes upon encoder movement like pot. I use clockwise for 10%, anticlockwise for -10% and switch for reset override to 100%. It works like a charm. I can share you my pi pico code if you wish to go this way.

**Reply by terjeio** [reply]
> I assume that all boards have an analog input.

3D printer boards has for the thermistor(s), most CNC boards likely not. Development boards are likely to have but are there free pins?

> The problem I see more in adding the value to the override.

See the [plugin code](https://github.com/grblHAL/Plugin_encoder/blob/9221b1a3cd8ec5064d3a9b9bef42e8c42b4fa423/encoder.c#L385-L394) for how I did it for encoder input.

The best way to implement for a pot is to write a plugin that install itself if there is an analog input available and then get the value via [hal.port.wait_on_input](http://svn.io-engineering.com/grblHAL/html/ioports_8h.html#ad33a7bae00e6224cb872c78a3d170f9c). Then write code for each driver (or board)  you want the plugin to run on that registers one or more analog inputs. Here is how I register an analog input with the HAL for the [iMXRT1062 driver](https://github.com/grblHAL/iMXRT1062/blob/7ee85521406e1b8146b1f14945bdfa47a936dfa2/grblHAL_Teensy4/src/ioports.c#L526-L547) (external I2C ADC) and for the [SAM3X8E driver](https://github.com/grblHAL/SAM3X8E/blob/ad1b7457430622415abc75b419e9d9d491c89cde/grblHAL_Due/src/ioports.c#L465-L468) (note: incomplete code).

FYI the [plasma plugin](https://github.com/grblHAL/Plugin_plasma/blob/master/thc.c) has code for reading analog input. Note that this is rather advanced as it modifies the HAL struct in order to add some virtual ports for M-command use (from gcode programs).

The alternative (and initially fastest?) is to write driver or board specific code - this is ok if for private use.
Or you can add driver code for encoder input and use encoder plugin.

@dresco Nice control! I am thinking about adding a tab to the Web Builder for 3rd party plugins. Yours could be a candidate for that?

**Reply by dresco** [reply]
> @dresco Nice control! I am thinking about adding a tab to the Web Builder for 3rd party plugins. Yours could be a candidate for that?

Yes absolutely. I think it will be a bit more useful once I have refined the hw implementation, but it can only motivate me! :smile:

**Reply by GrblGru** [reply]
@karoria
This is of course a good idea to use a little microcontroller as a collection point for operator activities.
The individual actions can then be sent as code via a 2nd USB interface to the PC.
I will test this in any case. Thanks for the good tip. :)

@terjeio
Thanks for your explanations. I have already looked at your source of the plugin.
To be honest, I have not yet fully understood how the plugins are integrated.
I still know too little of the GrblHAL system and have to read up on it now. This will take some time. 
Thanks for your great help.  :)

**Reply by terjeio** [reply]
> To be honest, I have not yet fully understood how the plugins are integrated.

Integration is via changing function pointers - think dynamic linking of code. A plugin typically only has one statically linked function, for initialization. The initialization function modifies whatever function pointers it wants to take over/link functionality into. There are a number of examples in the [Templates](https://github.com/grblHAL/Templates) repository, ranging from very simple ([LED blinker](https://github.com/grblHAL/Templates/blob/master/my_plugin/stm32F411blinky/my_plugin.c)) to advanced ([HPGL implementation](https://github.com/grblHAL/Templates/tree/master/my_plugin/hpgl)) - all these links into the core via `my_plugin_init()` which is a user reserved init function for plugins.
`my_plugin_init()` is declared as a [weak function](https://stackoverflow.com/questions/35507446/what-are-weak-functions-and-what-are-their-uses-i-am-using-a-stm32f429-micro-co) in the core and a user supplied version will be automatically called if found - so no need to change any existing code to activate the plugin.
